### DISCUSSION TITLE
Extending exisitng Architecture

### DISCUSSION THREAD

**Discussion by clark-ja** [Vector35/binaryninja-api#3230] [Architectures] [answered]
is it possible to extend pre-existing architectures e.g. via the python API? My use case would be to extend mips32 with the mips16e2 ase without having to re-model the entire underlying architecture...

**Comment by psifertex** [comment, marked as answer]
Yes, there are two ways. The first is using an architecture extension but that is useful for small changes to existing architectures: 

[https://github.com/Vector35/binaryninja-api/tree/dev/examples/x86_extension](https://github.com/Vector35/binaryninja-api/tree/dev/examples/x86_extension)

[https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/arch_hook.py](https://github.com/Vector35/binaryninja-api/blob/dev/python/examples/arch_hook.py)

The second method is to modify the existing open source architecture and either submit a PR or you can disable the built-in plugin and load your own in its place:

[https://github.com/Vector35/arch-mips](https://github.com/Vector35/arch-mips)

**Reply by clark-ja** [reply]
thanks for the quick reply. Just a little thing: the links in your reply are not correctly set and point all to the `arch-mips`...

**Reply by psifertex** [reply]
Where did you expect the link to go? I was intending to link to the arch repository which is where it goes when I test it.

**Reply by clark-ja** [reply]
You posted 3 links and all of them point to https://github.com/Vector35/arch-mips (as you can see when you hover your mouse over them). Not that that is a big issue... one can simply copy the text and paste it to the navigation bar...

**Reply by psifertex** [reply]
Whoops! Thanks, fixing now. I misunderstood and of course I tried the last link which worked as expected. 😉
